### PR TITLE
Updated the Best Practices for Symfony 4 and Flex

### PR DESCRIPTION
**Pull request by javiereguiluz**
Still WIP. Some chapters have not been reviewed yet.

### CODE REVIEW

File: best_practices/tests.rst [hunk 1]
```diff
@@ -48,14 +49,12 @@ A functional test can be as easy as this::
...
-                array('/blog/category/fixture-category'),
-                array('/archives'),
-                // ...
-            );
+            yield ['/'];
+            yield ['/posts'];
+            yield ['/post/fixture-post-1'];
+            yield ['/blog/category/fixture-category'];
+            yield ['/archives'];
+            yield [.];
```
Comment (yceruto) [inline best_practices/tests.rst]: `yield [.]` => `// ...`?

**Review by yceruto — COMMENTED** [review]
Nice! :) 

Probably the section "Template Locations" needs changes too.

**Comment by javiereguiluz** [comment]
This is now ready to review. Here is the list of added/removed/changed best practices:

### Added

* Use the Symfony Skeleton to create new Symfony-based projects.
* Use autowiring to automate the configuration of application services.
* Don't add the ``Action`` suffix to the methods of the controller actions.

### Removed

* Don't define parameters for the classes of your services.
* Create only one bundle called AppBundle for your application logic.

### Changed

```diff
-Use the Symfony Installer to create new Symfony-based projects.
+Use Composer and Symfony Flex to create and manage Symfony applications.
```

```diff
-Make your controller extend the FrameworkBundle base controller...
+Make your controller extend the ``AbstractController`` base controller...
```

```diff
-Store the translation files in the ``app/Resources/translations/`` directory.
+Store the translation files in the ``translations/`` directory at the root of your project.
```

```diff
-Store all your application's templates in ``app/Resources/views/`` directory.
+Store the application templates in the ``templates/`` directory at the root of your project.
```

```diff
-Define your Twig extensions in the ``AppBundle/Twig/`` directory.
+Define your Twig extensions in the ``src/Twig/`` directory.
```

```diff
-Store your assets in the ``public/`` directory.
+Store your assets in the ``assets/`` directory at the root of your project.
```

```diff
-Use Assetic to compile, combine and minimize web assets, unless you're
-comfortable with frontend tools like GruntJS.
+Use `Webpack Encore`_ to compile, combine and minimize web assets.
```

```diff
-For fine-grained restrictions, define a custom security voter;
-For restricting access to *any* object by *any* user via an admin
-interface, use the Symfony ACL.
+Define a custom security voter to implement fine-grained restrictions.
```

**Review by yceruto — APPROVED** [review]

File: best_practices/business-logic.rst [hunk 2]
```diff
@@ -25,15 +25,24 @@ Inside here, you can create whatever directories you want to organize things:
...
+.. best-practice::
 
-Let's create a new ``Slugger`` class inside ``src/Utils/`` and
-add the following ``slugify()`` method:
+    Use autowiring to automate the configuration of application services.
+
+:doc:`Service autowiring </service_container/autowiring>` is a feature provided
+by Symfony's Service Container to manage services with minimal configuration.
+It reads the type-hints on your constructor (or other methods) and automatically
+passes you the correct services. It can also add :doc:`service tags </service_container/tags>`
```
Comment (fabpot) [inline best_practices/business-logic.rst]: It passes the correct services to the methods, not 'you'.

File: best_practices/business-logic.rst [hunk 3]
```diff
@@ -42,30 +51,15 @@ add the following ``slugify()`` method:
...
-.. note::
-
-    If you're using the :ref:`default services.yml configuration <service-container-services-load-example>`,
-    the class is auto-registered as a service.
+If you're using the :ref:`default services.yaml configuration <service-container-services-load-example>`,
+this class is auto-registered as a service whose ID is ``App\Utils\Slugger`` (or
+simply ``Slugger::class`` if the class is already imported in your code).
 
 Traditionally, the naming convention for a service was a short, but unique
 snake case key - e.g. ``app.utils.slugger``. But for most services, you should now
```
Comment (fabpot) [inline best_practices/business-logic.rst]: I would remove this paragraph.

File: best_practices/creating-the-project.rst [hunk 4]
```diff
@@ -4,53 +4,45 @@ Creating the Project
...
 
-In the past, Symfony projects were created with `Composer`_, the dependency manager
-for PHP applications. However, the current recommendation is to use the **Symfony
-Installer**, which has to be installed before creating your first project.
+.. best-practice::
+
+    Use Composer and Symfony Flex to create and manage Symfony applications.
+
+`Composer`_ is the package manager used by modern PHP application to manage their
+dependencies. `Symfony Flex`_ is a Composer plugin designed to automatize some
```
Comment (fabpot) [inline best_practices/creating-the-project.rst]: automate?

File: best_practices/creating-the-project.rst [hunk 5]
```diff
@@ -4,53 +4,45 @@ Creating the Project
...
+project based on the most recent stable Symfony version available.
 
 .. tip::
 
-    Symfony releases are digitally signed for security reasons. If you want to
-    verify the integrity of your Symfony installation, take a look at the
-    `public checksums repository`_ and follow `these steps`_ to verify the
-    signatures.
+    The technical requirements to run Symfony are simple. If you want to check
+    if your system meets those requirements, read :doc:`/reference/requirements`.
```
Comment (fabpot) [inline best_practices/creating-the-project.rst]: What about recommending using the `req-checker` for that?

Comment (javiereguiluz) [reply best_practices/creating-the-project.rst]: Agree! The linked article (https://symfony.com/doc/master/reference/requirements.html) was revamped recently to remove everything and just explain how to use req-checker.

**Review by wouterj — APPROVED** [review]
Great stuff!

**Comment by weaverryan** [comment]
This is huge work Javier. THANK YOU!!!

**Pull request merged**